src/Smuggler/Plugin.hs

```diff
@@ -29,12 +44,67 @@ smugglerPlugin _ modSummary tcEnv = do
     let modulePath = ms_hspp_file modSummary
 
     fileContent <- liftIO $ BS.readFile modulePath
-    liftIO $ void $ hashStore cacheDir (smuggling modulePath) (takeFileName modulePath, fileContent)  -- TODO: remove takeFileName
+    uses <- readMutVar (tcg_used_gres tcEnv)
+    liftIO $ void $ hashStore cacheDir (smuggling uses modulePath) (takeFileName modulePath, fileContent)  -- TODO: remove takeFileName
 
     pure tcEnv
   where
-    smuggling :: FilePath -> ByteString -> IO ByteString
-    smuggling modulePath _ = do
+--    smuggling :: FilePath -> ByteString -> IO ByteString
+    smuggling uses modulePath _ = do
+        -- 1. find positions of unused imports
+        let user_imports = filter (not . ideclImplicit . unLoc) (tcg_rn_imports tcEnv)
+        let usage = findImportUsage user_imports uses
+        let unusedPositions = concatMap unusedLocs usage
+--        debugAST unusedPositions
+
+        -- 2. Remove positions of unused imports from annotations.
         (anns, ast) <- runParser modulePath  -- TODO: don't read file, use given ByteString
-        putStrLn $ exactPrint ast $ removeAnnAtLoc 4 31 anns
+        let purifiedAnnotations = foldl' (\ann (x, y) -> removeAnnAtLoc x y ann) anns unusedPositions
+        putStrLn $ exactPrint ast purifiedAnnotations
+
+        -- 3. Return empty ByteString
         pure ""
+
+unusedLocs ::ImportDeclUsage -> [(Int, Int)]
+unusedLocs (L (RealSrcSpan loc) decl, used, unused)
+    -- Do not warn for `import M ()`
+    | Just (False, L _ []) <- ideclHiding decl
+    = []
+
+    -- Note [Do not warn about Prelude hiding]
+    -- TODO: add ability to support custom prelude
+    | Just (True, L _ hides) <- ideclHiding decl
+    , not (null hides)
+    , pRELUDE_NAME == unLoc (ideclName decl)
+    = []
+
+    -- Nothing used; drop entire decl
+    -- TODO: drop every line of multiline import
+    -- TODO: optimize
+    | null used = map (srcSpanStartLine loc,) [1..100]
+
+    -- Everything imported is used; drop nothing
+    | null unused = []
+
+    -- only part of non-hiding import is used
+    | Just (False, L _ lies) <- ideclHiding decl
+    = unusedEntries lies
+
+    -- TODO: unused hidings
```

Comment:
Need to think on this one..

Reply:
Yeah, probably it will require to check all names inside module. On the other hand I remember that `hlint` does something similar...